src/crystal/system/win32/event_loop_iocp.cr

```diff
@@ -34,37 +35,62 @@ class Crystal::Iocp::EventLoop < Crystal::EventLoop
     iocp
   end
 
-  # Runs the event loop.
+  # Runs the event loop and enqueues the fiber for the next upcoming event or
+  # completion.
   def run_once : Nil
+    # Pull the next upcoming event from the event queue. This determines the
+    # timeout for waiting on the completion port.
+    # OPTIMIZE: Implement @queue as a priority queue in order to avoid this
+    # explicit search for the lowest value and dequeue more efficient.
     next_event = @queue.min_by(&.wake_at)
 
-    if next_event
-      now = Time.monotonic
+    unless next_event
+      Crystal::System.print_error "Warning: No runnables in scheduler. Exiting program.\n"
+      ::exit
+    end
```

Comment:
For the record, this cannot actually happen because `min_by` instead of `min_by?` is used

Reply:
nice catch!

Reply:
**thought (non blocking)**: is that the job of the eventloop to decide to exit? As stated in the error message, this is a matter for the scheduler.

libevent merely returns an error (no events), and the case is expected in my implementation for [RFC #2](https://github.com/crystal-lang/rfcs/pull/2): the schedulers use it to decide to start spinning then park themselves for example.

Reply:
Yeah, this is definitely not a final solution. But it's just a stop gap for some odd edge case. So far it has never happend because the branch couldn't be reached and you'd get a different error if the queue was empty.
I think this was introduced before we even had an actual event loop on Windows and things were much more brittle.